src/executor/graph_executor.cc

```diff
@@ -369,6 +606,314 @@ void GraphExecutor::Init(nnvm::Symbol symbol,
+ * to populate data entries of the graph. The created NDArrays
+ * for in_args, arg_grads and aux_states are passed to the
+ * front end to attach the created executor.
+ * In front end, if the simple_bind flow is trigger by
+ * _bind_ith_exec, the shared data arrays of DataParallelExecutorGroup
+ * and shared executor will be taken into account in creating
+ * NDArrays for in_args, arg_grads, and aux_states for resuing
+ * already allocated memory.
+ */
+void GraphExecutor::Init2(nnvm::Symbol symbol,
+                          const Context& default_ctx,
+                          const std::map<std::string, Context>& ctx_map,
+                          const std::vector<Context>& in_arg_ctxes,
+                          const std::vector<Context>& arg_grad_ctxes,
+                          const std::vector<Context>& aux_state_ctxes,
+                          const std::unordered_map<std::string, TShape>& arg_shape_map,
+                          const std::unordered_map<std::string, int>& arg_dtype_map,
+                          const std::vector<OpReqType>& grad_req_types,
+                          const std::unordered_set<std::string>& param_names,
+                          const std::vector<NDArray>& shared_exec_in_args,
+                          const std::vector<NDArray>& shared_exec_arg_grads,
+                          const std::vector<NDArray>& shared_exec_aux_states,
+                          std::vector<NDArray>* in_arg_vec,
+                          std::vector<NDArray>* arg_grad_vec,
+                          std::vector<NDArray>* aux_state_vec,
+                          std::unordered_map<std::string, NDArray>* shared_data_arrays,
+                          Executor* shared_exec) {
+  nnvm::Graph g = InitGraph2(symbol, default_ctx, ctx_map, in_arg_ctxes, arg_grad_ctxes,
+                             aux_state_ctxes, grad_req_types);
+  
+  // The following code of shape and dtype inferences and argument
+  // initialization is for simple_bind only. Regular bind operation
+  // should do this differently.
+
+  // Initialize arg_shapes and arg_dtypes for shape and type inferences.
+  // It contains all in_args and aux_states' shapes and types in a certain order.
+  const nnvm::IndexedGraph& idx = g.indexed_graph();
+  nnvm::ShapeVector arg_shapes(idx.input_nodes().size(), TShape());
+  nnvm::DTypeVector arg_dtypes(idx.input_nodes().size(), -1);
+  for (size_t i = 0; i < num_forward_inputs_; ++i) {
+    const uint32_t nid = idx.input_nodes().at(i);
+    const std::string& name = idx[nid].source->attrs.name;
+    auto it1 = arg_shape_map.find(name);
+    if (arg_shape_map.end() != it1) {
+      arg_shapes[i] = it1->second;
+    }
+    auto it2 = arg_dtype_map.find(name);
+    if (arg_dtype_map.end() != it2) {
+      arg_dtypes[i] = it2->second;
+    }
+  }
+  g = nnvm::pass::InferShape(g, arg_shapes, "__shape__");
+  g = nnvm::pass::InferType(g, arg_dtypes, "__dtype__");
```

Comment:
I don't quite understand the question. But the shape attr in the graph is designed as a vector. Its indices correspond to other data structures of the graph.